File: PandorasBox/Ahorn/entities/drivablecar.jl
module PandorasBoxIntroCar

using ..Ahorn, Maple

@mapdef Entity "pandorasBox/introCar" IntroCar(x::Integer, y::Integer, acceleration::Number=256.0, deceleration::Number=384.0, maxSpeed::Number=384.0, brokenDoor::Bool=false, nitroAcceleration::Number=448.0, nitroMaxDuration::Number=3.0, nitroRegenMultiplier::Number=0.2, facing::Integer=1)

const placements = Ahorn.PlacementDict(
    "Intro Car (Pandora's Box)" => Ahorn.EntityPlacement(
        IntroCar
    )
)

Ahorn.editingOptions(entity::IntroCar) = Dict{String, Any}(
    "facing" => Dict{String, Any}(
        "Left" => -1,
        "Right" => 1
    )
)

const bodySprite = "scenery/car/body"
const wheelsSprite = "scenery/car/wheels"

function Ahorn.selection(entity::IntroCar)
    x, y = Ahorn.position(entity)

    return Ahorn.coverRectangles([
        Ahorn.getSpriteRectangle(bodySprite, x, y, jx=0.5, jy=1.0),
        Ahorn.getSpriteRectangle(wheelsSprite, x, y, jx=0.5, jy=1.0),
    ])
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::IntroCar, room::Maple.Room)
    x, y = Ahorn.position(entity)

    scaleX = get(entity.data, "facing", 1)

    Ahorn.drawSprite(ctx, wheelsSprite, x, y, jx=0.5, jy=1.0, sx=scaleX, sy=1)
    Ahorn.drawSprite(ctx, bodySprite, x, y, jx=0.5, jy=1.0, sx=scaleX, sy=1)
end

end

File: PandorasBox/Loenn/entities/drivable_car.lua
local drawableSpriteStruct = require("structs.drawable_sprite")

local introCar = {}

local bodyTexture = "scenery/car/body"
local wheelsTexture = "scenery/car/wheels"

introCar.name = "pandorasBox/introCar"
introCar.placements = {
    name = "intro_car",
    data = {
        acceleration = 256.0,
        deceleration = 384.0,
        maxSpeed = 384.0,
        brokenDoor = false,
        keepCarSpeedOnExit = true,
        nitroAcceleration = 448.0,
        nitroMaxDuration = 3.0,
        nitroRegenMultiplier = 0.2,
        facing = 1
    }
}
introCar.fieldInformation = {
    facing = {
        fieldType = "integer"
    }
}

function introCar.sprite(room, entity)
    local sprites = {}

    local bodySprite = drawableSpriteStruct.fromTexture(bodyTexture, entity)
    bodySprite:setJustification(0.5, 1.0)
    bodySprite.depth = 1

    local wheelSprite = drawableSpriteStruct.fromTexture(wheelsTexture, entity)
    wheelSprite:setJustification(0.5, 1.0)
    wheelSprite.depth = 3

    table.insert(sprites, bodySprite)
    table.insert(sprites, wheelSprite)

    return sprites
end

return introCar